File: tests/runtime/flavorful/host.py
from typing import Tuple, List
from helpers import TestWasi
from flavorful import Flavorful, FlavorfulImports
import flavorful as e
from flavorful.imports import imports as i
from flavorful.types import Result, Ok, Err
import wasmtime

class MyImports:
    def f_list_in_record1(self, a: i.ListInRecord1) -> None:
        pass

    def f_list_in_record2(self) -> i.ListInRecord2:
        return i.ListInRecord2('list_in_record2')

    def f_list_in_record3(self, a: i.ListInRecord3) -> i.ListInRecord3:
        assert(a.a == 'list_in_record3 input')
        return i.ListInRecord3('list_in_record3 output')

    def f_list_in_record4(self, a: i.ListInAlias) -> i.ListInAlias:
        assert(a.a == 'input4')
        return i.ListInRecord4('result4')

    def f_list_in_variant1(self, a: i.ListInVariant1V1, b: i.ListInVariant1V2, c: i.ListInVariant1V3) -> None:
        assert(a == 'foo')
        assert(b == Err('bar'))
        assert(c == 'baz')

    def f_list_in_variant2(self) -> i.ListInVariant2:
        return 'list_in_variant2'

    def f_list_in_variant3(self, a: i.ListInVariant3) -> i.ListInVariant3:
        assert(a == 'input3')
        return 'output3'

    def errno_result(self) -> Result[None, i.MyErrno]:
        return Err(i.MyErrno.B)

    def list_typedefs(self, a: i.ListTypedef, c: i.ListTypedef3) -> Tuple[i.ListTypedef2, i.ListTypedef3]:
        assert(a == 'typedef1')
        assert(c == ['typedef2'])
        return (b'typedef3', ['typedef4'])

    def list_of_variants(self, a: List[bool], b: List[Result[None, None]], c: List[i.MyErrno]) -> Tuple[List[bool], List[Result[None, None]], List[i.MyErrno]]:
          assert(a == [True, False])
          assert(b == [Ok(None), Err(None)])
          assert(c == [i.MyErrno.SUCCESS, i.MyErrno.A])
          return (
                [False, True],
                [Err(None), Ok(None)],
                [i.MyErrno.A, i.MyErrno.B],
          )

def run() -> None:
    store = wasmtime.Store()
    wasm = Flavorful(store, FlavorfulImports(MyImports(), TestWasi()))

    wasm.test_imports(store)
    wasm.f_list_in_record1(store, e.ListInRecord1("list_in_record1"))
    assert(wasm.f_list_in_record2(store) == e.ListInRecord2(a="list_in_record2"))

    assert(wasm.f_list_in_record3(store, e.ListInRecord3("list_in_record3 input")).a == "list_in_record3 output")
    assert(wasm.f_list_in_record4(store, e.ListInRecord4("input4")).a == "result4")

    wasm.f_list_in_variant1(store, "foo", e.Err("bar"), 'baz')
    assert(wasm.f_list_in_variant2(store) == "list_in_variant2")
    assert(wasm.f_list_in_variant3(store, "input3") == "output3")

    assert(isinstance(wasm.errno_result(store), e.Err))

    r1, r2 = wasm.list_typedefs(store, "typedef1", ["typedef2"])
    assert(r1 == b'typedef3')
    assert(r2 == ['typedef4'])

if __name__ == '__main__':
    run()

File: tests/runtime/flavorful/host.rs
use anyhow::Result;

wit_bindgen_host_wasmtime_rust::generate!({
    import: "../../tests/runtime/flavorful/imports.wit",
    default: "../../tests/runtime/flavorful/exports.wit",
    tracing: true,
    name: "exports",
});

#[derive(Default)]
pub struct MyImports;

impl imports::Imports for MyImports {
    fn f_list_in_record1(&mut self, ty: imports::ListInRecord1) {
        assert_eq!(ty.a, "list_in_record1");
    }

    fn f_list_in_record2(&mut self) -> imports::ListInRecord2 {
        imports::ListInRecord2 {
            a: "list_in_record2".to_string(),
        }
    }

    fn f_list_in_record3(&mut self, a: imports::ListInRecord3) -> imports::ListInRecord3 {
        assert_eq!(a.a, "list_in_record3 input");
        imports::ListInRecord3 {
            a: "list_in_record3 output".to_string(),
        }
    }

    fn f_list_in_record4(&mut self, a: imports::ListInAlias) -> imports::ListInAlias {
        assert_eq!(a.a, "input4");
        imports::ListInRecord4 {
            a: "result4".to_string(),
        }
    }

    fn f_list_in_variant1(
        &mut self,
        a: imports::ListInVariant1V1,
        b: imports::ListInVariant1V2,
        c: imports::ListInVariant1V3,
    ) {
        assert_eq!(a.unwrap(), "foo");
        assert_eq!(b.unwrap_err(), "bar");
        match c {
            imports::ListInVariant1V3::String(s) => assert_eq!(s, "baz"),
            imports::ListInVariant1V3::F32(_) => panic!(),
        }
    }

    fn f_list_in_variant2(&mut self) -> Option<String> {
        Some("list_in_variant2".to_string())
    }

    fn f_list_in_variant3(&mut self, a: imports::ListInVariant3) -> Option<String> {
        assert_eq!(a.unwrap(), "input3");
        Some("output3".to_string())
    }

    fn errno_result(&mut self) -> Result<(), imports::MyErrno> {
        imports::MyErrno::A.to_string();
        format!("{:?}", imports::MyErrno::A);
        fn assert_error<T: std::error::Error>() {}
        assert_error::<imports::MyErrno>();
        Err(imports::MyErrno::B)
    }

    fn list_typedefs(
        &mut self,
        a: imports::ListTypedef,
        b: imports::ListTypedef3,
    ) -> (imports::ListTypedef2, imports::ListTypedef3) {
        assert_eq!(a, "typedef1");
        assert_eq!(b.len(), 1);
        assert_eq!(b[0], "typedef2");
        (b"typedef3".to_vec(), vec!["typedef4".to_string()])
    }

    fn list_of_variants(
        &mut self,
        bools: Vec<bool>,
        results: Vec<Result<(), ()>>,
        enums: Vec<imports::MyErrno>,
    ) -> (Vec<bool>, Vec<Result<(), ()>>, Vec<imports::MyErrno>) {
        assert_eq!(bools, [true, false]);
        assert_eq!(results, [Ok(()), Err(())]);
        assert_eq!(enums, [imports::MyErrno::Success, imports::MyErrno::A]);
        (
            vec![false, true],
            vec![Err(()), Ok(())],
            vec![imports::MyErrno::A, imports::MyErrno::B],
        )
    }
}

fn run(wasm: &str) -> Result<()> {
    let (exports, mut store) = crate::instantiate(
        wasm,
        |linker| imports::add_to_linker(linker, |cx| -> &mut MyImports { &mut cx.imports }),
        |store, module, linker| Exports::instantiate(store, module, linker),
    )?;

    exports.test_imports(&mut store)?;

    exports.f_list_in_record1(
        &mut store,
        ListInRecord1 {
            a: "list_in_record1",
        },
    )?;
    assert_eq!(exports.f_list_in_record2(&mut store)?.a, "list_in_record2");

    assert_eq!(
        exports
            .f_list_in_record3(
                &mut store,
                ListInRecord3Param {
                    a: "list_in_record3 input"
                }
            )?
            .a,
        "list_in_record3 output"
    );

    assert_eq!(
        exports
            .f_list_in_record4(&mut store, ListInAliasParam { a: "input4" })?
            .a,
        "result4"
    );

    exports.f_list_in_variant1(
        &mut store,
        Some("foo"),
        Err("bar"),
        ListInVariant1V3::String("baz"),
    )?;
    assert_eq!(
        exports.f_list_in_variant2(&mut store)?,
        Some("list_in_variant2".to_string())
    );
    assert_eq!(
        exports.f_list_in_variant3(&mut store, Some("input3"))?,
        Some("output3".to_string())
    );

    assert!(exports.errno_result(&mut store)?.is_err());
    MyErrno::A.to_string();
    format!("{:?}", MyErrno::A);
    fn assert_error<T: std::error::Error>() {}
    assert_error::<MyErrno>();

    let (a, b) = exports.list_typedefs(&mut store, "typedef1", &["typedef2"])?;
    assert_eq!(a, b"typedef3");
    assert_eq!(b.len(), 1);
    assert_eq!(b[0], "typedef4");
    Ok(())
}
